src/rail_gui.cpp

```diff
@@ -762,6 +762,47 @@ struct BuildRailToolbarWindow : Window {
 	static HotkeyList hotkeys;
 };
 
+/**
+ * Selects RailType based on _last_built_railtype and global modifiers from RailToolbarWidgetModifiers
+ *
+ * If GHK_MOD_* modifiers were used hotkey replaced with last selected widget
+ * or WID_RAT_AUTORAIL (if toolbar was closed).
+ *
+ * Without wrapping, so if you're on RAILTYPE_RAIL you'll stay there. Same for upper boundary.
+ *
+ * @param [out] hotkey Replaced by last user selected widget or WID_RAT_AUTORAIL if prev/next railtype used
+ * @return selected RailType
+ */
+static RailType ChooseRailTypeOnGlobalHotkey(int &hotkey)
+{
+	extern RailType _last_built_railtype;
+	RailType rt = _last_built_railtype;
+	bool change_rt = false;
+	if (HasBit(hotkey, GHK_MOD_RAT_PREV_RAILTYPE)) {
+		ClrBit(hotkey, GHK_MOD_RAT_PREV_RAILTYPE);
+		change_rt = true;
+		do {
+			rt--;
+		} while (rt > RAILTYPE_BEGIN && !HasRailtypeAvail(_local_company, rt));
+		// RailType is byte (unsigned char), so it overflows after 0
+		if (rt >= RAILTYPE_END) rt = _last_built_railtype;
```

Comment:
Can this underflow? The while loop uses `rt > RAILTYPE_BEGIN`, so I'd expect a minimum of RAILTYPE_BEGIN

Reply:
You're correct, it will stop at `RAILTYPE_BEGIN`. It seems I left this guard from older patch.

Reply:
Pushed updated & rebased patch